test/lit/passes/gufa-cast-all.wast

```diff
@@ -145,4 +174,85 @@
       (local.get $a)
     )
   )
+
+  ;; CHECK:      (func $strings-target (type $3) (param $a stringref) (param $b stringview_wtf8) (param $c stringview_wtf16) (param $d stringview_iter)
+  ;; CHECK-NEXT:  (drop
+  ;; CHECK-NEXT:   (ref.as_non_null
+  ;; CHECK-NEXT:    (local.get $a)
+  ;; CHECK-NEXT:   )
+  ;; CHECK-NEXT:  )
+  ;; CHECK-NEXT:  (drop
+  ;; CHECK-NEXT:   (ref.as_non_null
+  ;; CHECK-NEXT:    (local.get $b)
+  ;; CHECK-NEXT:   )
+  ;; CHECK-NEXT:  )
+  ;; CHECK-NEXT:  (drop
+  ;; CHECK-NEXT:   (ref.as_non_null
+  ;; CHECK-NEXT:    (local.get $c)
+  ;; CHECK-NEXT:   )
+  ;; CHECK-NEXT:  )
+  ;; CHECK-NEXT:  (drop
+  ;; CHECK-NEXT:   (ref.as_non_null
+  ;; CHECK-NEXT:    (local.get $d)
+  ;; CHECK-NEXT:   )
+  ;; CHECK-NEXT:  )
+  ;; CHECK-NEXT: )
+  (func $strings-target
+    (param $a (ref null string))
+    (param $b (ref null stringview_wtf8))
+    (param $c (ref null stringview_wtf16))
+    (param $d (ref null stringview_iter))
+    ;; These params are only called with non-null values, so we can cast them
+    ;; to non-null. We must not use ref.cast for that, which is disallowed on
+    ;; strings and views.
```

Comment:
Make these params `anyref` and we still have a problem, though, right? This PR is a good optimization improvement (although I think OptimizeInstructions makes the same improvement), but it doesn't really solve the underlying problem.

Reply:
I don't follow - what problem do you mean?

If you mean a problem with a string/view that flows into anyref, then that isn't the issue I observed in the fuzz bug, it was just this specific thing: that we emitted a `ref.cast` in order to force a string/view to be non-nullable. string/view flowing into an anyref is a separate issue though, if a codebase does that, of course.

Reply:
Wouldn't we still emit a `ref.cast` targeting string views if the params were anyref?

Reply:
Are you saying if we passed string views into anyref params? Yes, that can be a problem, as I agreed, but it doesn't occur in practice (yet?) unlike the problem this PR fixes.

Or, if I've misunderstood you, can you explain what you mean, maybe with a testcase?

Reply:
I'm not saying it happens in practice, but the fundamental bug here seems to be that we consider the string views to be subtypes of `any` while V8 does not. GUFA emitting `ref.cast` for them follows from that bug, and this PR works around that in a particular case, but not in general.

This PR seems fine, but I see it as partially addressing a symptom of the underlying bug, so we should still fix the underlying bug as well.

Another fix for the particular symptom would be to run OptimizeInstructions after GUFA.

Reply:
Yeah, I agree this works around the issue. But it does it in a useful way - even though this overlaps with OptimizeInstructions, it's a small addition, and it makes the output here more idiomatic (and, actually valid even without running OptimizeInstructions). I added a comment for that.

Reply:
(And I don't know how to fix the underlying issue - we were discussing it might need various internalize/externalize operations... but as it doesn't happen in practice, I hope we don't need it.)